### PR TITLE
post run introspection

### LINKED ISSUE #50: add documentation for capturing information about schema changes
we need to have some kind of clear message as to what changed (table, change, load_id, time) that we can pass to a function that will forward this somewhere. I can then add an example for slack.

### LINKED ISSUE #134: Schema change log access
We need to be able to access schema changes that have happened during a load. We should ideally be very informative, as the volume of schema events is low, so capturing all of them is not overdoing it.

We should capture information such as:
- Load id or timestamp
- namespace info: Dataset/schema, table, column name, metadata
- PIPELINE NAME (multiple pipelines could load to the same staging layer or tables)
- type of change: New column added, variant column added, (other?)

This information should be accessible from the pipeline after loading.

Documentation should be created or examples.

### LINKED ISSUE #184: post run pipeline state introspection
**Background**
Checking what happened with the data during pipeline `run` should be easier both in the Python code and cli. 

**Tasks**
1. * [x] document how to get state and trace via cli and code
2. * [x] document `delete_completed_jobs` in `load`. False by default. document what is left behind
3. * [x] change it to delete only completed jobs as it was the case before.
4. * [x] add method to pipeline remove completed package fully. 
5. * [x] load storage should return a status of the package with schema name, jobs, table names and statuses. failed packages should contain error names. also links to files should be present. the info should be json loadable with dlt #73 
6. * [x] load module should persist the effective schema updates: a list of tables and columns that were added/updated. it should include flags for variant columns. see #134 
7. * [x] document PipelineStepFailed and `__context__` exception chaining. document terminal vs. non terminal exceptions when loading. `dlt` does not retry the `run` or any other step
8. * [x] add a config parameter to fail loading on a terminal exception in load job

### LINKED ISSUE #187: pipeline runtime documentation
In order to explain to our users how to monitor and control `dlt` pipelines at runtime, the following must be added to docs
1. * [x]  logging: log levels, formats, json logging, logging metrics
2. * [x]  raising exceptions and retries, terminal and non terminal exception, `PipelineStepFailed` and its context. Exceptions during load.
3. * [x]  tracing (with sentry)
4. * [ ]  metrics (prometheus)
5. * [x]  describe what `pipeline.run` does inside
6. * [x] how to hack the slack messages

### PR DESCRIPTION
**Pull request by rudolfix**
closes #134 
closes #50 
closes #184 
closes #187

### CODE REVIEW

**Review by TyDunn — APPROVED** [review]
Docs look good now. Monster PR! 👹

**Review by burnash — CHANGES_REQUESTED** [review]
Nice work on the PR! I noticed a few issues and I'll take a closer look tomorrow and give you more detailed feedback.

File: dlt/common/runtime/slack.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import requests
+from dlt.common import json, logger
+
+
+def send_slack_message(incoming_hook: str, message: str, is_markdown: bool = True) -> None:
+    """Sends a `message` to  Slack `incoming_hook`, by default formatted as markdown. Returns http status code"""
```
Comment (burnash) [inline dlt/common/runtime/slack.py:6]: The docstring mentions that the function returns the HTTP status code, but the function is defined to return None.

File: dlt/common/runtime/slack.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    r = requests.post(incoming_hook,
+        data= json.dumps({
```
Comment (burnash) [inline dlt/common/runtime/slack.py:8]: There's an extra space between = and json. flake8 is likely not strict enough ;)

File: dlt/cli/pipeline_command.py [hunk 3]
```diff
@@ -31,23 +38,93 @@ def pipeline_command(operation: str, pipeline_name: str, pipelines_dir: str) ->
...
         completed_loads = p.list_completed_load_packages()
-        for load_id in completed_loads:
+        normalized_loads = p.list_normalized_load_packages()
+        for load_id in completed_loads + normalized_loads:  # type: ignore
             fmt.echo("Checking failed jobs in load id '%s'" % fmt.bold(load_id))
-            for job, failed_message in p.list_failed_jobs_in_package(load_id):
-                fmt.echo("JOB: %s" % fmt.bold(os.path.abspath(job)))
-                fmt.secho(failed_message, fg="red")
+            failed_jobs = p.list_failed_jobs_in_package(load_id)
+            if len(failed_jobs) > 0:
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply dlt/cli/pipeline_command.py:95]: you are right. and I just learned that you can have `else` for `for` loop
https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

File: dlt/cli/pipeline_command.py [hunk 4]
```diff
@@ -31,23 +38,93 @@ def pipeline_command(operation: str, pipeline_name: str, pipelines_dir: str) ->
...
+                for failed_job in p.list_failed_jobs_in_package(load_id):
+                    fmt.echo("JOB: %s(%s)" % (fmt.bold(failed_job.job_file_info.job_id()), fmt.bold(failed_job.job_file_info.table_name)))
+                    fmt.echo("JOB file type: %s" % fmt.bold(failed_job.job_file_info.file_format))
+                    fmt.echo("JOB file path: %s" % fmt.bold(failed_job.file_path))
+                    if verbosity > 0:
+                        fmt.echo(failed_job.asstr(verbosity))
+                    fmt.secho(failed_job.failed_message, fg="red")
+                    fmt.echo()
+                else:
+                    fmt.echo("No failed jobs found")
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply dlt/cli/pipeline_command.py:105]: you are right. and I just learned that you can have `else` for `for` loop
https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

**Pull request merged**